firedrake/parameters.py

```diff
@@ -4,29 +4,428 @@
+        :arg help: Help information for the key
+        :arg depends: Specify whether a key is dependent on another key, the depended key
+            must be in the same Parameters class and must be of BoolType
+        :arg visible_level: Visible level of the key, default to be 0"
+        """
+        if val_type is not None:
+            self._type = val_type
+        else:
+            self._type = UnsetType()
+        self.visible_level = visible_level
+        self.help = help
+        self.depends = depends
+
+    @property
+    def help(self):
+        """Help information for the key"""
+        try:
+            return self._help or "No help available"
+        except AttributeError:
+            return "No help available"
+
+    @help.setter
+    def help(self, help):
+        self._help = help
+
+    @property
+    def type(self):
+        """Type information for the key. Must be subclass of
+            :class:`firedrake.Parameter.KeyType`"""
+        return self._type
+
+    @type.setter
+    def type(self, new_type):
+        if isinstance(new_type, KeyType):
+            self._type = new_type
+        else:
+            raise ValueError(new_type + "is not a type!")
+
+    @property
+    def visible_level(self):
+        """Visible level of the key, default to be 0"""
+        return self._visible_level
+
+    @visible_level.setter
+    def visible_level(self, new_level):
+        self._visible_level = new_level
+
+    def validate(self, value):
+        """Validate a input value for current key"""
+        return self._type.validate(value)
+
+    def set_wrapper(self, callable):
+        """Set a wrapper of input value"""
+        self._wrapper = callable
+
+    def set_unwrapper(self, callable):
+        """Set an unwrapper for the input value for display"""
+        self._unwrapper = callable
+
+    def wrap(self, value):
+        """Wrap a value if wrapper is set, else return the value unmodified"""
+        if hasattr(self, "_wrapper"):
+            return self._wrapper(value)
+        else:
+            return value
+
+    def unwrap(self, value):
+        """Unwrap a value if unwrapper is set, else return the value unmodified"""
+        if hasattr(self, "_unwrapper"):
+            return self._unwrapper(value)
+        else:
+            return value
+
+    def __getstate__(self):
+        # Remove non-picklable wrapper and unwrapper functions
+        d = self.__dict__.copy()
+        if hasattr(self, "_wrapper"):
+            del d["_wrapper"]
+        if hasattr(self, "_unwrapper"):
+            del d["_unwrapper"]
+        return d
+
+    @property
+    def depends(self):
+        """Specify whether a key is dependent on another key, the depended key
+            must be in the same Parameters class and must be of BoolType"""
+        if hasattr(self, "_depends"):
+            return self._depends
+        else:
+            return None
+
+    @depends.setter
+    def depends(self, key):
+        self._depends = key
```

Comment:
Move type stuff to (say) `firedrake/config_types.py` (better names appreciated).